src/OpenFeature.Contrib.Providers.Flagd/FlagdProvider.cs

```diff
@@ -0,0 +1,375 @@
+using Google.Protobuf.WellKnownTypes;
+using Grpc.Net.Client;
+using OpenFeature.Model;
+
+using Schema.V1;
+using Value = OpenFeature.Model.Value;
+using ProtoValue = Google.Protobuf.WellKnownTypes.Value;
+
+namespace OpenFeature.Contrib.Providers.Flagd
+{
+    /// <summary>
+    ///     FlagdProvider is the OpenFeature provider for flagD.
+    /// </summary>
+    public sealed class FlagdProvider : FeatureProvider
+    {
+        private readonly Service.ServiceClient _client;
+        private readonly Metadata _providerMetadata = new Metadata("flagD Provider");
+
+        /// <summary>
+        ///     Constructor of the provider.
+        ///     <param name="url">The URL of the flagD server</param>
+        ///     <exception cref="ArgumentNullException">if no url is provided.</exception>
+        /// </summary>
+        public FlagdProvider(Uri url)
+        {
+            if (url == null)
+            {
+                throw new ArgumentNullException(nameof(url));
+            }
+
+            _client = new Service.ServiceClient(GrpcChannel.ForAddress(url));
+        }
+
+        // just for testing, internal but visible in tests
+        internal FlagdProvider(Service.ServiceClient client)
+        {
+            _client = client;
+        }
+
+        /// <summary>
+        /// Get the provider name.
+        /// </summary>
+        public static string GetProviderName()
+        {
+            return Api.Instance.GetProviderMetadata().Name;
+        }
+
+        /// <summary>
+        ///     Return the metadata associated to this provider.
+        /// </summary>
+        public override Metadata GetMetadata() => _providerMetadata;
+
+        /// <summary>
+        ///     ResolveBooleanValue resolve the value for a Boolean Flag.
+        /// </summary>
+        /// <param name="flagKey">Name of the flag</param>
+        /// <param name="defaultValue">Default value used in case of error.</param>
+        /// <param name="context">Context about the user</param>
+        /// <returns>A ResolutionDetails object containing the value of your flag</returns>
+        public override async Task<ResolutionDetails<bool>> ResolveBooleanValue(string flagKey, bool defaultValue, EvaluationContext context = null)
+        {
+            try
+            {
+                var resolveBooleanResponse = await _client.ResolveBooleanAsync(new ResolveBooleanRequest
+                {
+                    Context = ConvertToContext(context),
+                    FlagKey = flagKey
+                });
+
+                return new ResolutionDetails<bool>(
+                    flagKey: flagKey,
+                    value: resolveBooleanResponse.Value,
+                    reason: resolveBooleanResponse.Reason,
+                    variant: resolveBooleanResponse.Variant
+                );
+            }
+            catch (Grpc.Core.RpcException e)
+            {
+                return GetDefaultWithException<bool>(e, flagKey, defaultValue);
```

Comment:
I would simplify the logic here a bit just to throw an informative error (you still make need . The SDK itself will always handle any errors. The most straightforward thing for a provider to do is simply throw an error with the correct code. https://github.com/open-feature/dotnet-sdk/blob/main/src/OpenFeature/Error/FeatureProviderException.cs

This will be slightly easier when [this](https://github.com/open-feature/dotnet-sdk/issues/110) is done, but you can do it now by just setting the right error type on the existing generic error class.